libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,214 @@
+
+	return true;
+}
+
+}
+
+namespace solidity::frontend
+{
+
+
+shared_ptr<FunctionCallGraphBuilder::ContractCallGraph> const FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	m_contract = &_contract;
+
+	m_graph = make_shared<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::CreationRoot;
+	m_currentDispatch = SpecialNode::CreationDispatch;
+	visitConstructor(
+		_contract,
+		_contract.annotation().linearizedBaseContracts.cbegin() + 1,
+		_contract.annotation().linearizedBaseContracts.cend()
+	);
+	m_currentNode = SpecialNode::Unset;
+	m_currentDispatch = SpecialNode::RuntimeDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_graph->edges.count(funcDef))
+				visitCallable(funcDef);
+
+	m_contract = nullptr;
+	solAssert(!nodeSet(m_currentNode), "Current node not properly reset.");
+
+	return m_graph;
+}
+
+// Function not called right away -> dispatch
+// FunctionCall without ref -> dispatch
+//
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		if (*_identifier.annotation().requiredLookup == VirtualLookup::Virtual)
+			callable = &callable->resolveVirtual(*m_contract);
+
+		solAssert(
+			*_identifier.annotation().requiredLookup != VirtualLookup::Super,
+			"Only MemberAccess can have lookup 'super'"
+		);
+
+		// Ignore all calls that are not internal
+		if (auto funcType = dynamic_cast<FunctionType const*>(callable->type()))
+			if (funcType->kind() != FunctionType::Kind::Internal)
+				return true;
+
+		if (!m_graph->edges.count(callable))
+		{
+			// Create edge to creation dispatch
+			if (!_identifier.annotation().calledDirectly)
+				add(m_currentDispatch, callable);
+			visitCallable(callable);
+		}
+
+		solAssert(nodeSet(m_currentNode), "");
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(MemberAccess const& _memberAccess)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
+	{
+		if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+		{
+			CallableDeclaration const* callableNode = nullptr;
+
+			if (auto node = std::get_if<ASTNode const*>(&m_currentNode))
+				callableNode = dynamic_cast<CallableDeclaration const*>(*node);
+
+			ContractDefinition const* super =
```

Comment:
Please extract `CompilerContext::superFunction` to a common place instead of re-implementing it here.